nemo/collections/asr/models/sortformer_diar_models.py

```diff
@@ -430,21 +531,280 @@ def forward(
+            [spkcache, fifo, chunk_pre_encode_embs], [spkcache_lengths, fifo_lengths, chunk_pre_encode_lengths]
+        )
+
+        # encode the concatenated embeddings
+        spkcache_fifo_chunk_fc_encoder_embs, spkcache_fifo_chunk_fc_encoder_lengths = self.frontend_encoder(
+            processed_signal=spkcache_fifo_chunk_pre_encode_embs,
+            processed_signal_length=spkcache_fifo_chunk_pre_encode_lengths,
+            bypass_pre_encode=True,
+        )
+
+        # forward pass for inference
+        spkcache_fifo_chunk_preds = self.forward_infer(
+            spkcache_fifo_chunk_fc_encoder_embs, spkcache_fifo_chunk_fc_encoder_lengths
+        )
+        return spkcache_fifo_chunk_preds, chunk_pre_encode_embs, chunk_pre_encode_lengths
+
+    def forward_streaming(
+        self,
+        processed_signal,
+        processed_signal_length,
+    ):
+        """
+        The main forward pass for diarization inference in streaming mode.
+
+        Args:
+            processed_signal (torch.Tensor): tensor containing audio waveform
+                Dimension: (batch_size, num_samples)
+            processed_signal_length (torch.Tensor): tensor containing lengths of audio waveforms
+                Dimension: (batch_size,)
+
+        Returns:
+            total_pred (torch.Tensor): tensor containing predicted speaker labels for the current chunk
+                and all previous chunks
+                Dimension: (batch_size, pred_len, num_speakers)
+        """
+        streaming_state = self.sortformer_modules.init_streaming_state(
+            batch_size=processed_signal.shape[0], async_streaming=self.async_streaming, device=self.device
+        )
+
+        batch_size, ch, sig_length = processed_signal.shape
+        processed_signal_offset = torch.zeros((batch_size,), dtype=torch.long, device=self.device)
+
+        if dist.is_available() and dist.is_initialized():
+            local_tensor = torch.tensor([sig_length], device=processed_signal.device)
+            dist.all_reduce(
+                local_tensor, op=dist.ReduceOp.MAX, async_op=False
+            )  # get max feature length across all GPUs
+            max_n_frames = local_tensor.item()
+            if dist.get_rank() == 0:
+                logging.info(f"Maximum feature length across all GPUs: {max_n_frames}")
+        else:
+            max_n_frames = sig_length
+
+        if sig_length < max_n_frames:  # need padding to have the same feature length for all GPUs
+            pad_tensor = torch.full(
+                (batch_size, ch, max_n_frames - sig_length),
+                self.negative_init_val,
+                dtype=processed_signal.dtype,
+                device=processed_signal.device,
+            )
+            processed_signal = torch.cat([processed_signal, pad_tensor], dim=2)
+
+        att_mod = False
+        if self.training:
+            r = random.random()
+            if r < self.sortformer_modules.causal_attn_rate:
```

Comment:
Adding a quick comment here explaining causal_attn_rate, causal_attn_rc and why this is required would be beneficial

Reply:
This needs to be revisited when we add documentations and example codes